src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs

```diff
@@ -239,5 +339,223 @@ private CspParameters GetPrivateKeyCsp()
+                keyProvInfo.dwFlags = machineKey ? CryptAcquireContextFlags.CRYPT_MACHINE_KEYSET : 0;
+                keyProvInfo.dwKeySpec = keySpec;
+
+                if (!Interop.crypt32.CertSetCertificateContextProperty(
+                    pal._certContext,
+                    CertContextPropId.CERT_KEY_PROV_INFO_PROP_ID,
+                    CertSetPropertyFlags.None,
+                    &keyProvInfo))
+                {
+                    pal.Dispose();
+                    throw Marshal.GetLastWin32Error().ToCryptographicException();
+                }
+            }
+
+            return pal;
+        }
+
+        private static int GuessKeySpec(
+            CngProvider provider,
+            string keyName,
+            bool machineKey,
+            CngAlgorithmGroup algorithmGroup)
+        {
+            if (provider == CngProvider.MicrosoftSoftwareKeyStorageProvider ||
+                provider == CngProvider.MicrosoftSmartCardKeyStorageProvider)
+            {
+                // Well-known CNG providers, keySpec is 0.
+                return 0;
+            }
+
+            const int NTE_BAD_KEYSET = unchecked((int)0x80090016);
+
+            try
+            {
+                CngKeyOpenOptions options = machineKey ? CngKeyOpenOptions.MachineKey : CngKeyOpenOptions.None;
+
+                using (CngKey.Open(keyName, provider, options))
+                {
+                    // It opened with keySpec 0, so use keySpec 0.
+                    return 0;
+                }
+            }
+            catch (CryptographicException e)
+            {
+                Debug.Assert(
+                    e.HResult == NTE_BAD_KEYSET,
+                    $"CngKey.Open had unexpected error: 0x{e.HResult:X8}: {e.Message}");
+
+                CspParameters cspParameters = new CspParameters
+                {
+                    ProviderName = provider.Provider,
+                    KeyContainerName = keyName,
+                    Flags = CspProviderFlags.UseExistingKey,
+                    KeyNumber = (int)KeyNumber.Signature,
+                };
+
+                if (machineKey)
+                {
+                    cspParameters.Flags |= CspProviderFlags.UseMachineKeyStore;
+                }
+
+                if (algorithmGroup == CngAlgorithmGroup.Rsa)
+                {
+                    // Try the AT_SIGNATURE spot in each of the 4 RSA provider type values,
+                    // ideally one of them will work.
+                    const int PROV_RSA_FULL = 1;
+                    const int PROV_RSA_SIG = 2;
+                    const int PROV_RSA_SCHANNEL = 12;
+                    const int PROV_RSA_AES = 24;
+
+                    // These are ordered in terms of perceived likeliness, given that the key
+                    // is AT_SIGNATURE.
+                    int[] provTypes =
+                    {
+                        PROV_RSA_FULL,
+                        PROV_RSA_AES,
```

Comment:
Doesn't matter; all we're looking for is proof that the key uses AT_SIGNATURE.  Given that it's an AT_SIGNATURE key, RSA_FULL seems more likely.